### PR TITLE
Made maps actual hash-maps instead of linear arrays

### PR DESCRIPTION
**Pull request by AWoloszyn**
This should let us have giant maps without performance penalty.
There are a couple of things to investigate still. The primary one
is how to handle rehashing, we should probably be doubling the
size of the hash map instead of adding a fixed amount, that way when we
get really large arrays, we don't have constant rehashing operations,
which are significantly more expensive than the dynamic array approach.

### CODE REVIEW

File: core/codegen/arithmetic.go [hunk 1]
```diff
@@ -205,6 +210,23 @@ func (b *Builder) Div(x, y *Value) *Value {
 	}
 }
 
+// Div returns x % y. The types of the two values must be equal.
```
Comment (ben-clayton) [inline core/codegen/arithmetic.go:213]: Div / Rem ?

File: gapil/compiler/map.go [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+	u64Type := c.targetType(semantic.Uint64Type)
+	u32Type := c.targetType(semantic.Uint32Type)
+	if keyType != value.Type() {
+		fail("hashValue must be called with the given type, %+v, %+v", keyType, value.Type())
+	}
+
+	switch t := semantic.Underlying(t).(type) {
+	case *semantic.Builtin:
+		switch t {
+		case semantic.BoolType:
```
Comment (ben-clayton) [inline gapil/compiler/map.go:150]: Nit can be condensed down with:
```go
case semantic.BoolType, semantic.IntType, semantic.UintType ...
```

File: gapil/compiler/map.go [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+//   2) rehash if necessary
+//   3) find first bucket >= h, where used != full (mod capacity)
+//   4) Insert there, mark used = full
+
+// TODO: Investigate alternative rehash rule.
+//     Right now it is new_capacity = capacity + 16, likely we want new_capacity = capacity << 1
+// TODO: Investigate rehashing once #full + #previously_full > 80%
+//     If we end up with lots of insertions/deletions, this will prevent linear search
+
+const emptyBucket = uint64(0)
```
Comment (ben-clayton) [inline gapil/compiler/map.go:67]: Maybe:

```go
const (
    empty = iota
    full
    used
)
```

File: gapil/compiler/map.go [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+// struct Map {
+//    uint64_t count;
+//    uint64_t capacity;
+//    element* elements;
+// }
+
+// In order to look up something in the map
+// 1) h = Hashed Key % capacity
+// 2) Check element[h].
+//  If used == empty, then the element does not exist in the map
```
Comment (ben-clayton) [inline gapil/compiler/map.go:45]: Keep with names used below (`empty`, `used`, `full`)?

**Review by ben-clayton — APPROVED** [review]

**Comment by AWoloszyn** [comment]
Github refuses to update my other MR.
https://github.com/ben-clayton/gapid/pull/27

**Pull request closed without merging**